src/filters.cpp

```diff
@@ -157,9 +157,36 @@ NAN_METHOD(Filters::New) {
                     // insert the key/value into filters map
                     // TODO(dane): what if we have duplicate source-layer filters?
 
+                    // handle property array
+                    const v8::Local<v8::Value> layer_properties = layer->Get(Nan::New("properties").ToLocalChecked());
+                    if (layer_properties->IsNull() || layer_properties->IsUndefined()) {
+                        Nan::ThrowError("Property-Filters is not properly constructed.");
+                        return;
+                    }
+
+                    // NOTICE: If a layer is styled, but does not have a property, the property value will equal []
+                    // NOTICE: If a property is true, that means we need to keep all the properties
+                    filter_properties property;
+                    if (layer_properties->IsArray()) {
+                        // const auto propertyArray = layer_properties.As<v8::Array>();
+                        v8::Handle<v8::Array> propertyArray = v8::Handle<v8::Array>::Cast(layer_properties);
+                        uint32_t propertiesLength = propertyArray->Length();
+                        std::vector<std::string> values(propertiesLength);
```

Comment:
This is more efficient and safe:

```
std::vector<std::string> values;
values.reserve(propertiesLength);
for (uint32_t index = 0; index < propertiesLength; ++index) {
     Nan::Utf8String utf8_value(property_value);
     int utf8_len = utf8_value.length();
     if (utf8_len > 0) {
         values.emplace_back(*utf8_value, static_cast<std::size_t>(utf8_len));
     }
}
```

Reply:
Thanks, @springmeyer, I will make some change for this, but I still get some questions about this. 


1. Is this a risk: we reserve length to values, but we do not push anything into it if the value's length is 0.
Is that possible for traverse values and get the unpredictable value(if we reserve the length to 8, and we are not pushing eight values)?

-------
Questions need to research for @zmofei :
- [ ] `*v8::String::Utf8Value(property_value->ToString());` VS `Nan::Utf8String utf8_value`, how can we know the `utf8_value` method?
- [ ] `values.emplace_back(*utf8_value, static_cast<std::size_t>(utf8_len));` for the 2nd param what dose `static_cast<std::size_t>(utf8_len)` do and why we need pass a length to it?

Reply:
> if we reserve the length to 8, and we are not pushing eight values

It is not a risk since the only thing `values.reserve` does is reserve memory. So, if we don't end up adding 8 the only problem is a little bit more memory that might be requested than used. However, a good rule of thumb is to only use `values.reserve` when we know that 100% of the time it will add up to 8 (or the `propertiesLength`). In this case I think it always will unless the developer has passed in an empty string in the properties. So, perhaps we should throw an error in this case?

Reply:
> how can we know the utf8_value method?

https://github.com/nodejs/nan/blob/master/doc/v8_misc.md#api_nan_utf8_string

> for the 2nd param what dose static_cast<std::size_t>(utf8_len) do and why we need pass a length to it?

This is tricky. So what is happening is that `emplace_back()` "forwards" the arguments to the constructor of the type inside the `vector`.  So the `Args&&` at https://en.cppreference.com/w/cpp/container/vector/emplace_back basically says "pass the arguments to the next function". That "next" function is the constructor for `std::string`. Now the `std::string` constructor has a lot of different signatures, but this one matches `string::string(charT const* s, size_type count)`: we are passing the `data` as the first arg and its `size` as the second (https://en.cppreference.com/w/cpp/string/basic_string/basic_string)

Reply:
In this case that is the only difference: when you know the length of the string then it makes sense to be explicit and use it. One other case where it could make a difference is if the ‘const char *’ data is not null terminated such that the length is required in order to know where the string ends. However I think in this case we don’t absolutely need to pass the length since I presume the data is null terminated.